lib/Runtime/Library/DataView.h

```diff
@@ -104,106 +104,91 @@ namespace Js
 
         template<typename TypeName, typename PointerAccessTypeName>
-        Var GetValueWithCheck(uint32 byteOffset, BOOL isLittleEndian, const char16* funcName)
+        Var GetValueWithCheck(Var offset, BOOL isLittleEndian, const char16* funcName)
         {
             ScriptContext* scriptContext = GetScriptContext();
             if (this->GetArrayBuffer()->IsDetached())
             {
                 JavascriptError::ThrowTypeError(scriptContext, JSERR_DetachedTypedArray, funcName);
             }
-            if ((byteOffset + sizeof(TypeName) <= GetLength()) && (byteOffset <= GetLength()))
+            uint32 byteOffset = ArrayBuffer::ToIndex(offset, JSERR_DataView_InvalidOffset, scriptContext, GetLength() - sizeof(TypeName), false);
+
+            TypeName item;
+            TypeName *typedBuffer = (TypeName*)(buffer + byteOffset);
+            if (!isLittleEndian)
             {
-                TypeName item;
-                TypeName *typedBuffer = (TypeName*)(buffer + byteOffset);
-                if (!isLittleEndian)
-                {
-                    SwapRoutine<TypeName>(typedBuffer, &item);
-                }
-                else
-                {
-                    item = *static_cast<PointerAccessTypeName>(typedBuffer);
-                }
-                return JavascriptNumber::ToVarWithCheck(item, GetScriptContext());
+                SwapRoutine<TypeName>(typedBuffer, &item);
             }
             else
             {
-                JavascriptError::ThrowTypeError(scriptContext, JSERR_DataView_InvalidOffset);
+                item = *static_cast<PointerAccessTypeName>(typedBuffer);
             }
+            return JavascriptNumber::ToVarWithCheck(item, GetScriptContext());
         }
 
         template<typename TypeName>
-        inline void SetValue(uint32 byteOffset, TypeName value, const char16 *funcName, BOOL isLittleEndian = FALSE)
+        inline void SetValue(Var offset, TypeName value, const char16 *funcName, BOOL isLittleEndian = FALSE)
         {
-            SetValue<TypeName, TypeName*>(byteOffset, value, isLittleEndian, funcName);
+            SetValue<TypeName, TypeName*>(offset, value, isLittleEndian, funcName);
         }
 
         template<typename TypeName, typename PointerAccessTypeName>
-        void SetValue(uint32 byteOffset, TypeName value, BOOL isLittleEndian, const char16 *funcName)
+        void SetValue(Var offset, TypeName value, BOOL isLittleEndian, const char16 *funcName)
         {
             ScriptContext* scriptContext = GetScriptContext();
             if (this->GetArrayBuffer()->IsDetached())
             {
                 JavascriptError::ThrowTypeError(scriptContext, JSERR_DetachedTypedArray, funcName);
             }
-            if ((byteOffset + sizeof(TypeName) <= GetLength()) && (byteOffset <= GetLength()))
+            uint32 byteOffset = ArrayBuffer::ToIndex(offset, JSERR_DataView_InvalidOffset, scriptContext, GetLength() - sizeof(TypeName), false);
```

Comment:
>GetLength() - sizeof(TypeName) [](start = 106, length = 30)

Seems like we risk an underflow here if length is very small. Let's add a test case that makes sure we throw if the length is zero and we request element zero.

Reply:
May need to specifically check for that actually good shout.